craft_cli/messages.py

```diff
@@ -134,14 +137,53 @@ def _write_line(self, message: _MessageInfo) -> None:
         else:
             self.unfinished_stream = message.stream
 
+    def _write_bar(self, message: _MessageInfo) -> None:
+        """Write a progress bar to the screen."""
+        if self.prv_msg is None or self.prv_msg.end_line:
+            # first message, or previous message completed the line: start clean
+            maybe_cr = ""
+        elif self.prv_msg.ephemeral:
+            # the last one was ephemeral, overwrite it
+            maybe_cr = "\r"
+        else:
+            # complete the previous line, leaving that message ok
+            maybe_cr = ""
+            print(flush=True, file=self.prv_msg.stream)
+
+        numerical_progress = f"{message.bar_progress}/{message.bar_total}"
+        bar_percentage = min(message.bar_progress / message.bar_total, 1)  # type: ignore
+
+        # terminal size minus the text and numerical progress, and 5 (the cursor at the end,
+        # two spaces before and after the bar, and two surrounding brackets)
+        terminal_width = _get_terminal_width()
+        bar_width = terminal_width - len(message.text) - len(numerical_progress) - 5
+
+        # only show the bar with progress if there is enough space, otherwise just the
+        # message (truncated, if needed)
+        if bar_width > 0:
+            completed_width = math.floor(bar_width * min(bar_percentage, 100))
+            completed_bar = "#" * completed_width
```

Comment:
We could consider changing to a block character at some point in the future.

Reply:
Yes. @sergiusens already mentioned it. I just didn't know which character to use. Do you have one that we know to be in the different fonts that are used in terminals both in Linux and Windows?

Reply:
This should work https://github.com/verigak/progress/blob/master/progress/bar.py#L52, from this post, looks like the ones mentioned in this table https://mike42.me/blog/2018-06-make-better-cli-progress-bars-with-unicode-block-characters would work too

Reply:
I'd use the one that you mentioned first. It's unicode name is 'FULL BLOCK'... looks simple enough, it should be in all fonts (who know!).

We can leave "sub char resolution" for the future, IMO. I just opened #9 to not forget about it.

Reply:
Pushed the version using the FULL BLOCK.